drivers: can: expand return values reported by the CAN timing functions

henrikbrixandersen opened the pull request:
Expand the error return values reported by the CAN timing calculation functions to be able to distinguish between an out of range bitrate/sample point, an unsupported bitrate, and a resulting sample point outside the guard limit.

Move the check for the maximum allowed CAN bitrate to the corresponding can_calc_timing()/can_calc_timing_data() function.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

warasilapm commented on include/zephyr/drivers/can.h:
```diff
@@ -723,7 +723,8 @@ static inline const struct can_timing *z_impl_can_get_timing_max(const struct de
  * @param sample_pnt Sampling point in permill of the entire bit time.
```
-EINVAL seems most natural for arguments which are out of range rather than an invalid combination of arguments, but it's use here keeps legacy behavior there's no good error number which describes "a solution could not be found" in the error number list. -ERANGE or -ENOTSUP  may be reasonable for "no solution within the acceptable parameters", but the latter is already in use for "a solution could be found but not implemented by the specific CAN controller" so -ERANGE could be a descriptive addition here.

henrikbrixandersen replied:
Thank you for reviewing. Indeed, this is only the first step towards more detailed error codes from the CAN bitrate calculation functions.

My plan is to use `-EDOM` for parameters that exceed their domain (e.g. a sample point at 0% or 100%, or a too high bitrate). That is what is implemented by this PR.

The next step is to return `-ENOTSUP` for bitrates not supported by the CAN controller (if the CAN clock cannot be prescaled to provide the exact bitrate requested) and `-ERANGE` if the resulting sample point is outside the allowed window (+/- 5% of the requested sample point value). One could argue that `-ENOTSUP` could be replaced by `-EINVAL` but the argument value is really invalid, it's just not supported by the given CAN controller.

These return codes seem appropriate to me according to the comments of the minimal libc `errno.h` file:
- `#define EDOM 33         /**< Argument too large */`
- `#define ENOTSUP 134         /**< Unsupported value */`
- `#define ERANGE 34       /**< Result too large */`

The Linux `errno.3` man page (https://linux.die.net/man/3/errno) has these definitions for the same errno values:
- EDOM: Mathematics argument out of domain of function (POSIX.1, C99)
- ENOTSUP: Operation not supported (POSIX.1)
- ERANGE: Result too large (POSIX.1, C99)

henrikbrixandersen replied:
Another option could be to return `-EINVAL` for bitrates and sample points out-of-range (and remove `-EDOM` altogether), `-ENOTSUP` for bitrates without an integer match, and `-ERANGE` for calculated sample points exceeding the guard limit?

henrikbrixandersen replied:
I think I may actually prefer this over my current proposal.

warasilapm replied:
> Another option could be to return `-EINVAL` for bitrates and sample points out-of-range (and remove `-EDOM` altogether), `-ENOTSUP` for bitrates without an integer match, and `-ERANGE` for calculated sample points exceeding the guard limit?

> I think I may actually prefer this over my current proposal.

I'm not sure I actually like this mapping better now that I'm seeing the bigger picture. Reading the POSIX `-EDOM` description, though, `-EDOM` isn't a bad way to describe that the desired input parameters are outside of the acceptable range, even if it does alias the description of `-EINVAL` a bit. I always think of `-EINVAL`'s base case as checking for pointers set to `NULL` or incorrect types/etc.

On a slightly different note, is it worth it to differentiate between the integer match not existing and the controller not supporting a given bitrate? It intuitively makes sense to me that an integer solution for the segments exists but the controller wouldn't work properly. As an end user it may provide value when looking for the root cause of the problem, but maybe it's not actually useful.

henrikbrixandersen replied:
> I'm not sure I actually like this mapping better now that I'm seeing the bigger picture. Reading the POSIX `-EDOM` description, though, `-EDOM` isn't a bad way to describe that the desired input parameters are outside of the acceptable range, even if it does alias the description of `-EINVAL` a bit. I always think of `-EINVAL`'s base case as checking for pointers set to `NULL` or incorrect types/etc.

The two are kind of overlapping, but I too tend to view `-EINVAL` as truly invalid parameters, not just out-of-range parameters.

> On a slightly different note, is it worth it to differentiate between the integer match not existing and the controller not supporting a given bitrate? It intuitively makes sense to me that an integer solution for the segments exists but the controller wouldn't work properly. As an end user it may provide value when looking for the root cause of the problem, but maybe it's not actually useful.

What is the difference between the two? The CAN timing calculation function only has access to the CAN core clock and the allowed prescaler range, so if an integer match can be found using the two, the controller is assumed to support the bitrate. How else could a CAN controller not support a given bitrate?

warasilapm replied:
Okay I've thought some more and may have justifiable values:

- `-EINVAL` is used for invalid parameter checking, like for bit rates and sample points out of range. This error number is often used for parameter errors like this even though.
- `-EDOM` describes the situation where a solution can't be found between a core clock and the prescaler range. This follows because the desired inputs are not in the domain for the desired output. (Am I stretching the definition here? Maybe.)
- `-ERANGE` specifically describes the output sample point being beyond the guard limit. This differentiates between the speed being unreachable and there being no sample point solution. `-ERANGE` is usually supposed to be reserved for large number solutions outside of the range of the return type, but I think it fits well enough here.
- `-ENOTSUP` describes any situation where there may be a mathematic solution but some hardware feature prevents this from being implemented (perhaps the bitrate is too low, as suggested below).

The justifications may not be perfect I think this may give enough options to expand the error numbers returned while still mostly fitting their categories.

henrikbrixandersen replied:
I like it. However, I don't think we can reliably distinguish between a bitrate where no integer solution is found for the CAN core clock/prescaler range and a bitrate not supported by the CAN controller for some other reason - and I am not sure we need to either.

Here's why:
- `can_calc_timing*()` checks the bitrate against the limit for the given phase (arbitration or data), bit otherwise only has access to the CAN core clock and the prescaler range. If an integer solution is found for the requested bitrate, that function (internally `can_calc_timing_int()) has to assume that the CAN controller supports the given bitrate. It needs to be able to signal:
  - A requested bitrate being too high for the given phase (`-EINVAL`).
  - A requested sample point being out of range - either 0, or equal to or above 100% (`-EINVAL`)
  - Not being able to obtain the CAN core clock frequency (error reported by `can_get_core_clock()` should be propagated - it would be nice to limit the outcome here, but I'd much rather propagate the error code directly, as it may originate in a much lower layer such as a clock control driver).
  - A requested bitrate without an integer solution between CAN core clock and prescaler (either `-EDOM` or `-ENOTSUP` as discussed here)
  - A calculated sample point outside the guard limit (`-ERANGE`).
- `can_set_bitrate*()` checks if the bitrate within the limit of the CAN controller/transceiver combination, calls `can_calc_timing*()` before applying the calculated timing via `can_set_timing*()`. Now, `can_set_timing*()` calls into the driver API for the actual driver instance, which can then validate the timing parameters and check for "valid" but "unsupported" values (too low bitrates or otherwise unsupported bitrates are likely not among these checks unless the CAN controller does not support all values of prescaler range). It needs to be able to signal:
  - A bitrate above the limit of the CAN controller/transceiver combination (`-ENOTSUP`).
  - All the return values listed under `can_calc_timing*()` above.
  - The return value for `can_set_timing*()` (currently only `-EIO` is allowed per the API documentation).

As can be seen from the above, `-ENOTSUP` is used for indicating an unsupported bitrate in another situation already. I believe having to distinguish between _why_ a requested bitrate is not supported will only muddy the waters.

warasilapm replied:
This is all looking good. If it were solely up to me, I'd probably lean towards using `-EDOM` for bitrates that do not have a solution given the core clock and prescaler just to give more granularity. However, it also makes perfect sense to use `-ENOTSUP` in line with the checks done in `can_set_bitrate*()`. If that makes more sense to you as the maintainer, that's what should be done.

warasilapm commented on drivers/can/can_common.c:
```diff
@@ -88,10 +88,8 @@ static int can_calc_timing_int(uint32_t core_clock, struct can_timing *res,
-	    (!IS_ENABLED(CONFIG_CAN_FD_MODE) && bitrate > 1000000) ||
-	     (IS_ENABLED(CONFIG_CAN_FD_MODE) && bitrate > 8000000)) {
-		return -EINVAL;
+	if (bitrate == 0 || sp >= 1000) {
+		return -EDOM;
```
-EDOM seems reasonably descriptive for bitrates and samples points which are too large specifically, but not as much for the bitrate being zero. This probably doesn't matter given the return contract it fulfills.

henrikbrixandersen replied:
How come? A bitrate of 0 is outside of the allowed value domain.

warasilapm replied:
Yes, I am suggesting to use -EINVAL.

warasilapm commented on tests/drivers/can/timing/src/main.c:
```diff
@@ -172,7 +172,8 @@ static void test_timing_values(const struct device *dev, const struct can_timing
 	int err;
 
 	printk("testing bitrate %u, sample point %u.%u%% (%s): ",
-		test->bitrate, test->sp / 10, test->sp % 10, test->invalid ? "invalid" : "valid");
```
Does it make sense for the tests to _also_ test for -EINVAL separately from -EDOM? It looks like the else block of this conditional would provide that coverage but it's not explicit so I'm just checking here.

henrikbrixandersen replied:
No CAN controller should return `-EINVAL` here as it is today, but one of the ideas behind a more fine-grained error code result from the CAN bitrate APIs is to be able to test all CiA recommended bitrates and simply ignore the ones where the _bitrate_ is not supported by the given CAN controller (quite a few CAN controllers do not support the low frequency bitrates).

warasilapm reviewed (commented):
If there's an opportunity and need to expand the error numbers reported from these api functions, I'd personally been more granular and maybe updated the strange use of -EINVAL. However, this seems functional and reasonable to me.

henrikbrixandersen committed:
drivers: can: move bitrate limit checks to arbitration/data functions

Move the check for the maximum allowed CAN bitrate to the corresponding
can_calc_timing()/can_calc_timing_data() function.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: can: expand return values reported by the CAN timing functions

Expand the error return values reported by the CAN timing calculation
functions to be able to distinguish between an an out of range
bitrate/sample point, an unsupported bitrate, and a resulting sample point
outside the guard limit.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
@warasilapm I have updated this PR based on your feedback and our discussions. Please take another look.

warasilapm reviewed (approved):
Looks good to me.
